homeassistant/components/plugwise/sensor.py

```diff
@@ -0,0 +1,411 @@
+                            model,
+                        )
+                    )
+                else:
+                    entities.append(
+                        PwThermostatSensor(
+                            api,
+                            coordinator,
+                            entity["name"],
+                            dev_id,
+                            sensor,
+                            sensor_type,
+                        )
+                    )
+                _LOGGER.info("Added sensor.%s", entity["name"])
+
+        if single_thermostat is False:
+            for state in INDICATE_ACTIVE_LOCAL_DEVICE:
+                if state in data:
+                    entities.append(
+                        PwThermostatSensor(
+                            api,
+                            coordinator,
+                            entity["name"],
+                            dev_id,
+                            DEVICE_STATE,
+                            None,
+                        )
+                    )
+                    _LOGGER.info("Added sensor.%s_state", "{}".format(entity["name"]))
+                    break
+
+    async_add_entities(entities, True)
+
+
+class SmileSensor(SmileGateway):
+    """Represent Smile Sensors."""
+
+    def __init__(self, api, coordinator):
+        """Initialise the sensor."""
+        super().__init__(api, coordinator)
+
+        self._dev_class = None
+        self._state = None
+        self._unit_of_measurement = None
+
+    @property
+    def device_class(self):
+        """Device class of this entity."""
+        if not self._dev_class:
+            return None
+        return self._dev_class
+
+    @property
+    def state(self):
+        """Device class of this entity."""
+        if not self._state:
+            return None
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        if not self._unit_of_measurement:
+            return None
+        return self._unit_of_measurement
+
+
+class PwThermostatSensor(SmileSensor, Entity):
+    """Thermostat (or generic) sensor entities."""
+
+    def __init__(self, api, coordinator, name, dev_id, sensor, sensor_type):
+        """Set up the Plugwise API."""
+        super().__init__(api, coordinator)
+
+        self._api = api
+        self._gateway_id = self._api.gateway_id
+        self._dev_id = dev_id
+        self._sensor_type = sensor_type
+        self._entity_name = name
+        self._sensor = sensor
+        self._state = None
+        self._model = None
+        self._unit_of_measurement = None
+        if self._sensor_type is not None:
+            self._model = self._sensor_type[0]
+            self._unit_of_measurement = self._sensor_type[1]
+            self._dev_class = self._sensor_type[2]
+            self._icon = self._sensor_type[3]
+        else:
+            self._dev_class = "auxiliary"
+
```

Comment:
Does it make sense to break this up into two sensor classes here.  It looks like this class is supported two different data types.

Reply:
Actually the else should go, as otherwise we would introduce a new device_class (as per your previous comment). Towards the reason for the if-statement: devices that are not 'plugwise' but are disclosed by it fall under auxiliary.

Reply:
We do have one special sensor for showing the status (idle/heating/cooling) of the heating/cooling (auxiliary) device when there are more than one master-thermostats present.
This sensor is quite different from the other sensors. It would make sense to give that one it's own class.

Reply:
@bdraco there wouldn't be much difference (and some repeat code) if we take this approach, @bouwew already took some time to [work](https://github.com/plugwise/plugwise-beta/commit/9eb3c7c8253142fe7e32346832a4acdded1af937) on this - it will however nicely split the amount of `def _process_data` statements.

Reply:
The complexity reduction in `def _process_data` looks like a nice win.

Reply:
Committed that part, will need some side-by-side testing from both @bouwew and myself tomorrow morning (.eu TZ)

Reply:
I have a working version now. Will be added to this PR later today.